pkg/users/group-create-dialog.js

```diff
@@ -0,0 +1,169 @@
+            </FormGroup>
+        </Form>
+    );
+}
+
+function validate_name(name, groups) {
+    if (!name)
+        return _("No group name specified");
+
+    for (let i = 0; i < name.length; i++) {
+        if (!is_valid_char_name(name[i]))
+            return _("The group name can only consist of letters from a-z, digits, dots, dashes and underscores.");
+    }
+
+    for (let k = 0; k < groups.length; k++) {
+        if (groups[k].name == name)
+            return _("A group with this name already exists");
+    }
+
+    return null;
+}
+
+function validate_group(id, groups) {
+    if (!id)
+        return _("No ID specified");
+
+    const id_num = parseInt(id);
+    if (id_num.toString() !== id || id_num < 0)
+        return _("The group ID must be positive integer.");
+
+    for (let k = 0; k < groups.length; k++) {
+        if (groups[k].gid == id_num)
+            return _("A group with this ID already exists");
+    }
+
+    return null;
+}
+
+export function group_create_dialog(groups, setGroupsCardExpanded) {
+    cockpit.file("/etc/login.defs").read()
+            .then(logindef => {
+                const gids = groups
+                        .filter(g => g.name !== 'nobody')
+                        .map(group => group.gid);
+                const min_gid = parseInt(logindef.match(/^GID_MIN\s+(\d+)/m)[1] || 100);
+                const max_gid = parseInt(logindef.match(/^GID_MAX\s+(\d+)/m)[1] || 65534);
+
+                change("id", (Math.max(min_gid, Math.max(...gids.filter(id => id != max_gid)) + 1) + 1).toString());
+            });
+
+    let dlg = null;
+    const state = {
+        name: "",
+        id: "",
+    };
+    let errors = { };
+
+    function change(field, value) {
+        state[field] = value;
+        errors = { };
+
+        update();
+    }
+
+    function validate(name, id) {
+        const errs = { };
+
+        errs.name = validate_name(name, groups);
+        errs.id = validate_group(id, groups);
+        errors = errs;
+
+        return !has_errors(errs);
+    }
+
+    function create(name, id) {
+        const valid = validate(name, id);
+        if (valid) {
+            const group_add_cmd = ["/usr/sbin/groupadd", name, "-g", id];
+
+            return cockpit.spawn(group_add_cmd, { superuser: "require", err: "message" });
+        } else {
+            update();
+            return Promise.reject();
+        }
+    }
+
+    function update() {
+        const props = {
+            id: "groups-create-dialog",
+            title: _("Create new group"),
+            body: <GroupCreateBody state={state} errors={errors} change={change} />
+        };
+
+        const footer = {
+            actions: [
+                {
+                    caption: _("Create"),
+                    style: "primary",
+                    clicked: () => {
+                        return create(state.name, state.id).then(() => setGroupsCardExpanded(true));
```

Comment:
Where are cockpit.spawn() errors handled? I don't see a place where the error would be shown in a dialog?

Reply:
It's in the cockpit-components-dialog component internally.

Reply:
I chose to use that component because the rest of the accounts code uses it as well. I see the problems you see with readability though

Reply:
Thanks! If it works, it's fine. Resolving.